src/python/pants/util/contextutil.py

```diff
@@ -393,3 +402,23 @@ def with_overwritten_file_content(file_path):
   finally:
     with open(file_path, 'w') as f:
       f.write(file_original_content)
+
+
+@contextmanager
+def http_server(handler_class):
+  def serve(port_queue, shutdown_queue):
+    httpd = TCPServer(("", 0), handler_class)
+    httpd.timeout = 0.1
+    port_queue.put(httpd.server_address[1])
+    while shutdown_queue.empty():
+      httpd.handle_request()
+
+  port_queue = Queue()
+  shutdown_queue = Queue()
+  t = threading.Thread(target=lambda: serve(port_queue, shutdown_queue))
+  t.start()
+
+  yield port_queue.get(block=True)
+
+  shutdown_queue.put(True)
+  t.join()
```

Comment:
This should likely be in a `finally:` block to ensure that it runs in case of an exception.